1038 Archiving and restoring a node DB (#1137)
casper-network/docs: source/docs/casper/operators/advanced-topics/archiving-and-restoring.md

# Archiving and Restoring a Database

> This article describes processes for the compression and decompression of a Casper node database and streaming from a backup location.

Reviewer: ```suggestion
This documentation describes processes for the compression and decompression of a Casper node database and streaming from a backup location.
```
Article just feels a little strange.

Here are some examples of a Mainnet DB compression at block 741160:

| Level   | Time (min:sec)  | Size    |
|---------|-----------------|---------|
| 12      | 29:20           | 15.8 GB |
| 15      | 46:15           | 13.0 GB |
| 17      | 87:42           | 13.0 GB |
| 19      | 197:08          | 12.9 GB |

For local backups, using 1-5 is a great compression speed-to-size trade-off.  

### Thread count

The `-T[thread count]` is the number of threads that `zstd` should use for compression. If running a script or command on varying machines, use `T0` to allow `zstd` to detect the number of cores and run with the same number of threads as the detected cores. A speed-up can be obtained for machines with multiple threads per core by configuring a thread count near the number of threads. It is advisable to stay within the number of CPU threads. The recommendations in this article will use `-T0`.

### Long-distance matching

> The `--long=31` argument is where we see the most space gained by the algorithm because it controls the size of the matching window in powers of 2 (2**31 is 2 GB). The downside is that it requires 2.0 GB memory during compression and decompression as it looks ahead and rebuilds ahead. The default is 27 or 128 MB.

Reviewer: ```suggestion
The `--long=31` argument is where we see the most space gained by the algorithm because it controls the size of the matching window in powers of 2 (2**31 is 2 GB). The downside is that it requires 2.0 GB memory during compression and decompression as it looks and rebuilds ahead. The default is 27 or 128 MB.
```

At compression 19, we see a 30 GB file using the default 128 MB look ahead, and a 13 GB file using 2 GB look ahead. Since all validators should have 16-32 GB of memory, we keep this at `--long=31`.

> An important note is that decompression requires a compatible argument. Trying with a different long-distance matching value will throw an error, but the value to be used will be given.

Reviewer: ```suggestion
An important note is that decompression requires a compatible argument. Trying with a different long-distance matching value will result in an error. However, it will also return the necessary value to provide.
```

### Summary of commands

The general command for compression is:

```bash
tar -b 4096 -cv --sparse . | zstd -15 -cv -T0 --long=31 > [path_to]/file.tar.zst
```

For local backups, use a lower compression level:

```bash
tar -b 4096 -cv --sparse . | zstd -5 -cv -T0 --long=31 > [path_to]/file.tar.zst
```

## Decompression

`zstd -d` is the command for decompression; however, the same `--long` value used for compression must be specified. For all `casper-node` DB-related decompression, you will likely use this command:

```bash
zstd -cd --long=31 <.tar.zst file>
```

If `--long=31` is omitted, you might see an error such as this, which also gives you the solution:

```bash
./casper.tar.zst : Decoding error (36) : Frame requires too much memory for decoding 
./casper.tar.zst : Window size larger than maximum : 2147483648 > 134217728
./casper.tar.zst : Use --long=31 or --memory=2048MB
```

> Pipe the `zstd` result into a `tar -xv` command. Also, create the decompressed files using `sudo -u casper`, because the files will be used by the `casper-node`. Run the following command inside an empty DB location:

Reviewer: ```suggestion
You can then use the `zstd` result to populate a `tar -xv` command. Also, create the decompressed files using `sudo -u casper`, because the files will be used by the `casper-node`. Run the following command inside an empty DB location:
```

I'm not sure how universal the usage of "Pipe" is here.

```bash
zstd -cd --long=31 <.tar.zst file> | sudo -u casper tar -xv
```

To fix ownership, use this command:

```bash
sudo /etc/casper/node_util.py fix_permissions
```

## Streamed Decompression

If a `.tar.zst` archive is hosted on a website and you will not need the file after decompressing, you can stream it into the process using `curl`, which can output to stdout with `--output` and stream binary to your terminal.

```bash
curl -s --output - <URL for tar.zstd file>
```

> If you pipe the output into the previous process, you can decompress the files from `curl` directly into a local directory:

Reviewer: ```suggestion
If you use the output along with the previous process, you can decompress the files from `curl` directly into a local directory:
```
